### PR TITLE
Unidirectional Streams

### LINKED ISSUE #176: RST_STREAM breaks HPACK
@martinthomson noticed an issue in the QPACK draft that's equally true with HPACK in -01:

If you RST_STREAM a stream with a missing QUIC STREAM frame that contains an HPACK frame (HEADERS, PUSH_PROMISE), that missing frame never gets retransmitted.  That, in turn, means that no future HPACK frame can be interpreted, ever, because you're missing part of your history.

Oops.

My first inclination is to change the HTTP draft to say that message control streams MUST NOT be reset.  If you don't want the message body, reset the data stream, but you MUST read the message headers.  If, for some reason, you cannot do so, then terminate the connection.

Anyone see a barrier to that (other than the price of ginormous headers)?

### LINKED ISSUE #441: fix priority grouping node ambiguity
in h2 'phantom' (unopened) streams can be used for grouping nodes for priority purposes. This is simultaneously very useful and very concerning from a resource commitment purpose. I'm aware of some servers looking to limit their use through implicit resource allocations, and I'm aware of some clients looking to use them more agressively (i.e. a grouping node per tab).

Perhaps we should at least make these explicitly negotiated so they can be cleaned up, limited, etc..?

### LINKED ISSUE #515: A server can't send SETTINGS first
Stream 1 is a client-initiated stream.  That means that the server can't send its SETTINGS frame to the client until it reads the end of the client handshake and stream 1 is opened.  That's annoying because the server could otherwise send frames to the client after it sends its ServerHello.

In addition to SETTINGS, the server might use this otherwise dead air to send things like ORIGIN frames or secondary certificates.

This would not be a problem if streams were unidirectional ;)

### PR DESCRIPTION
**Pull request by martinthomson**
## Transport Changes

Streams are unidirectional.  Each has three states: idle, open, and closed.  I separated the transitions for sending and receiving because that turned out to be easier to explain.  The reordering thing makes them different in subtle ways.

That's all.  The changes in transport are relatively small and they simplify streams a lot.  That it also makes the transport more generic is a nice bonus.

## HTTP Changes

This is where the bulk of the changes are.

### Stream Correlation

Previously request and response were implicitly correlated, as was the data correlated with the request or response headers.  With this change, that correlation each stream has a header that explicitly correlates these.

There are 5 types of stream: connection control, request, response, data, and push.  The first two have a simple header that has a type.  The next two reference another stream in their header, which ties request to response and data to request or response.  Push streams each reference a PUSH_PROMISE using a newly minted push ID (I decided that was cleaner than what I proposed at the interim).

I chose backward references rather than forward references for two reasons.  First, request and response correlation can't use forward references because that would mean the client would be exerting (uncoordinated) control over server streams.  Second, that gives the server the most flexibility in terms of how it answers requests (see #281).  The cost of using backward references is that endpoints have to check that the backward references aren't bad, either referencing the wrong type of stream, or with multiple references to the same stream.

The presence or absence of a message body is signaled after the initial header block using a HAS_BODY frame.  This empty frame indicates that another stream will include the body of the message - or a promise for a body.  I would like to eliminate this stream split.  See #245 and #557 for more details on that, though we need to fix #176 first, which brings us back to QPACK/QCRAM again.

### Prioritization

This changes prioritization so that it identifies requests.  I only made the minimal changes here, which means that #441 doesn't get fixed at the same time, I've left that for later (see below).

### Cancelling Pushes

Because server push doesn't create streams with PUSH_PROMISE, I had to create a way to cancel them between the time that the PUSH_PROMISE is sent and when the push stream is created.  That's called RST_PUSH.

Update: the latest version changes RST_PUSH to CANCEL_REQUEST which allows a client to cancel both requests and server pushes.

## Things That Need Improvement

I haven't based this on #171.  I think that functionality is good, but the last time I looked at the PR I didn't like some of the changes Mike made there.  (That's a taste thing.)  This really assumes that we accept something very much like #171.

This really needs QPACK/QCRAM.  Right now, it is impossible to cancel some types of request using RST_STREAM because not all messages will have a body (#176 again).  On balance, given that bodies are what you really want to kill, it's not disastrous, but it's a problem nonetheless.  I think that we all agree that this is something that we need to fix, but we haven't reached the point where we agree on the details of the fix.

## Things That Want Improvement

This also really wants headers and data on the same stream .  It doesn't exactly need that, but merging the streams would make things a lot easy, both conceptually and structurally.

I chose the simplest possible encoding for all of the fields that I touched.  That means that stream IDs are all 32 bits in size and the HAS_BODY frame takes an entire 9 octets.  For things that are that common, there are many ways in which byte efficiency could be improved.

The push ID changes suggest a fix for #441.  Use of an as-yet-not-created ID as a node in the priority tree would allow for prioritization to use "empty" nodes.  Needs more thought.

# Fixes

Closes #515, #240, #281, #175.

(MikeBishop:  Edited to remove occurrences of "fix <issue num>" that were saying this PR doesn't fix that issue.  GitHub was getting confused.)

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
-peer-initiated streams that are "idle" if there is loss or reordering of
-packets.  Receiving these frames also causes the stream to become "open".
-
-An endpoint MUST NOT send a STREAM or RST_STREAM frame for a stream ID that is
-higher than the peers advertised maximum stream ID (see
-{{frame-max-stream-id}}).
-
+Sending a STREAM or RST_STREAM frame causes the identified stream to become
+"open" for a sending endpoint.  New streams use the next stream available
+identifier, as described in {{stream-id}}.  An endpoint MUST NOT send a STREAM
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: next available stream identifier?

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
+order.
 
+From the "open" state, the sending endpoint causes the stream to become "closed"
+by sending a RST_STREAM frame, or by sending a STREAM frame with the FIN flag
+set after sending all preceding data.
 
-### half-closed (remote) {#state-hc-remote}
+A stream becomes "closed" at the receiving endpoint when all data is received
+and a STREAM frame with a FIN flag is received or when a RST_STREAM frame is
+received.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: It is not clear if data must be send/received up to the RST_STREAM offset.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
-Reordering might cause frames to be received after closing, see
-{{state-hc-remote}}.
+An endpoint could continue receiving frames for the stream even after the stream
+is closed for sending if packets are reordered.  At the receiving endpoint,
+retransmissions of data in STREAM frames, or retransmissions of STREAM_BLOCKED
+and RST_STREAM frames might arrive after the stream becomes "closed".  At the
+sending endpoint, MAX_STREAM_DATA frames might be received after the stream
+becomes "closed".  Frames received on a "closed" stream SHOULD be discarded.  An
+endpoint MAY choose to limit the period over which it ignores frames and treat
+frames that arrive after this time as being in error.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: Perhaps a note on fatal connection error if receiving frames at later offset than close - otherwise a transmitter might abuse flow control limits - though the benefit of that is unclear. Perhaps also a note here stating the both RST_STREAM and FIN defines a specific end offset - even if discussed in more detail in specific frames.

**Comment by mikkelfj** [comment]
I can't speak for HTTP/2 that I have not studied in detail, but I believe this is a very significant improvement for the transport layer.

As mentioned above, DISINTEREST (#171) or some other mechanism for receiver to cancel is needed, and because it can cause some race conditions on stream state, I think it would be good to consider in more detail before settling on a unidirectional solution.

I am concerned about the STREAM_BLOCKED being able to force the receivers stream in open state if it is above the stream limit, or did I miss something. I know it is also there to hint at a need for more streams, but I think that should be an application level issue. Regardless - if naively implemented, a transmitter could force a lot of undesired resource usage at receiver by allocating stream state. I suggest that any stream related transmission above MAX_STREAM_LIMIT is a fatal error.

As to MAX_STREAM_LIMIT - this was a great improvement to avoid race conditions in the bi-directional design, and I do immediately suggest removing it. However, I think it might be worthwhile to consider alternatives for a better insight into the problem. For example: assume that we do not require incremental stream identifiers: there is no internal gain in incremental order such as keeping a simple stream table because old streams can randomly linger (e.g. stream 0), so open streams must be tracked via a map anyway. With a unidirectional design there is no ambigouty of the senders view of how many streams are at most open at the receiving end. It is the difference between received ACKS for RST_STREAM and FIN STREAMS and streams placed in senders open state.

Unless there is a flaw in the above argument, this suggests that streams can be opened with arbitrary identifiers. If we permit that, applications could use their own identifiers. However, it becomes intractable to make sure that a stream identifier is not being reused. Thus, the next question is: is it a problem to reuse stream identifiers once they have been closed? A sender cannot start reusing the stream before receiving an ACK for RST_STREAM or FIN STREAM of the old version of the stream and in this case the receiver cannot have many doubts about a new stream arriving being new. However, there may still be spurious retransmissions. This can be managed by looking at packet numbers but requires to remember the packet number of a closed stream, which is complicated.

Thus, I believe the MAX_STREAM_LIMIT is still the simplest approach.

**Comment by martinthomson** [comment]
You can't open a stream that exceeds the value your peer sent in MAX_STREAM_ID, STREAM_BLOCKED included.

@mikkelfj, your suggestion to allow use of arbitrary stream identifiers is counter to something we've discussed at length.  The current design allows an endpoint to maintain a map of open streams.  Then the state of other streams is easily determined.  Streams lower than the largest open stream ID are either in the map or are closed, streams with higher numbers are idle.

This is why I think that "fixing" #634 is not a good idea because it means additional state to track closed vs. idle on streams below the largest open stream ID.  And because there is nothing to be gained from opening out of order.  That's even more true with with this change, which removes any vestigial semantics attached to stream identifiers.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Seems clear to me: 
1. when all data is received and a STREAM frame with a FIN flag is received, or
2. when a RST_STREAM frame is received.

Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: It'ts not clear to me if the semantics are that the receiver can expect data to be delivered up the RST_FRAME offset, or the transmitter can just decide to cancel at that offset whether data before that was ACK'ed. The formulation of the RST_STREAM suggests the receiver should not wait around, but RST_STREAM could also be used to send half a file and use RST_STREAM instead of FIN, even if this is probably not the design intent.

**Comment by mikkelfj** [comment]
> @mikkelfj, your suggestion to allow use of arbitrary stream identifiers

I would call it a suggestion - I just wanted to ensure the previous discussion still holds up in the new configuration, and based on what I wrote, and what you have added, it seems to be the case.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: The formulation "all data received" on the FIN flag, implies that will not be the case for RST_STREAM, so I agree, it is covered in a terse form.

File: draft-ietf-quic-http.md [hunk 4]
```diff
@@ -322,37 +460,43 @@ is successfully established, the proxy sends a HEADERS frame containing a 2xx
...
-this priority scheme, a given stream can be designated as dependent upon another
-stream, which expresses the preference that the latter stream (the "parent"
-stream) be allocated resources before the former stream (the "dependent"
-stream). Taken together, the dependencies across all streams in a connection
-form a dependency tree. The structure of the dependency tree changes as PRIORITY
-frames add, remove, or change the dependency links between streams.
+this priority scheme, a given request can be designated as dependent upon
+another request, which expresses the preference that the latter stream (the
+"parent" request) be allocated resources before the former stream (the
+"dependent" request). Taken together, the dependencies across all request in a
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: nit: the dependencies across all request**s**

File: draft-ietf-quic-http.md [hunk 5]
```diff
@@ -322,37 +460,43 @@ is successfully established, the proxy sends a HEADERS frame containing a 2xx
...
+connection form a dependency tree. The structure of the dependency tree changes
+as PRIORITY frames add, remove, or change the dependency links between request.
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: nit: the dependency links between request**s**

File: draft-ietf-quic-http.md [hunk 6]
```diff
@@ -180,82 +180,221 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+A message that contains a payload body includes a HAS_BODY frame immediately
+after the first header block.  This indicates that another stream will include
+the payload body.  A data stream is opened to carry the payload body.  The data
+stream MUST be closed immediately after sending the body.  If the message does
+not contain a body, a data stream MUST NOT be opened and the HAS_BODY message
+MUST NOT be sent.  The "chunked" transfer encoding defined in Section 4.1 of
+{{!RFC7230}} MUST NOT be used.
 
 Trailing header fields are carried in an additional header block on the message
 control stream. Such a header block is a sequence of HEADERS frames with End
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: Sending trailing header fields on the control stream seems like a hangover from earlier design.

File: draft-ietf-quic-http.md [hunk 7]
```diff
@@ -363,23 +507,25 @@ pushes via the SETTINGS_DISABLE_PUSH setting in the SETTINGS frame (see
...
 
-The server push response is conveyed in the same way as a non-server-push
-response, with response headers and (if present) trailers carried by HEADERS
-frames sent on the control stream, and response body (if any) sent via the
-corresponding data stream.
+The server push response is conveyed on a push stream.  Aside from the stream
+header, which identifies the response stream and PUSH_PROMISE frame, a push
+stream is identical to a regular response stream ({{stream-response}}), with
+response headers and (if present) trailers carried by HEADERS frames sent on the
+control stream, and response body (if any) indicated with a HAS_BODY frame and
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: My reading of the design is that trailing header fields would be carried on the push stream, not the control stream.

File: draft-ietf-quic-http.md [hunk 8]
```diff
@@ -401,6 +547,14 @@ All frames have the following format:
 
 ## Frame Definitions {#frames}
 
+### HAS_BODY {#frame-has-body}
+
+The HAS_BODY frame indicates that the HTTP message contains a payload body.  It
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: Is it missing a frame type?

File: draft-ietf-quic-http.md [hunk 9]
```diff
@@ -712,6 +936,21 @@ HTTP_MULTIPLE_SETTINGS (0x10):
...
+: A stream header contained an unknown type or referenced an invalid stream.
+
+HTTP_UNMATCHED_RESPONSE (0x13):
+: A response stream referenced an invalid stream.
+
+HTTP_UNMATCHED_DATA (0x14):
+: A data stream referenced an invalid stream.
+
+HTTP_UNMATCHED_PUSH (0x15):
+: A push stream referenced an server push request.
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: nit: **a** server push request

File: draft-ietf-quic-http.md [hunk 10]
```diff
@@ -758,8 +997,8 @@ QUIC simply by replacing Stream 0 in HTTP/2 with Stream 1 in HTTP/QUIC.
 Below is a listing of how each HTTP/2 frame type is mapped:
 
 DATA (0x0):
-: Instead of DATA frames, HTTP/QUIC uses a separate data stream.  See
-  {{stream-mapping}}.
+: Instead of DATA frames, HTTP/QUIC uses a separate data stream.  The code point
+  for DATA frames is used for the HAS_BODY frame. See {{frame-has-body}}.
```
Comment (LPardue) [inline draft-ietf-quic-http.md]: Maybe this answers my earlier question about HAS_BODY type?

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -2755,10 +2722,10 @@ stream.  For a stream that is reset, the final offset is carried explicitly in
 the RST_STREAM frame.  Otherwise, the final offset is the offset of the end of
 the data carried in STREAM frame marked with a FIN flag.
 
-An endpoint will know the final offset for a stream when the stream enters the
-"half-closed (remote)" state.  However, if there is reordering or loss, an
-endpoint might learn the final offset prior to entering this state if it is
-carried on a STREAM frame.
+An endpoint will know the final offset for a stream when it receives a STREAM
+frame with a FIN flag or a RST_STREAM frame.  If there is reordering or loss,
+an endpoint receive a STREAM frame with a FIN flag prior to the stream entering
```
Comment (LPardue) [inline draft-ietf-quic-transport.md]: Is a word is missing here?

**Comment by mnot** [comment]
**Reminder** -- substantive discussion / feedback needs to happen on issues and/or the mailing list, not on PRs. Editorial-level discussion only, please.

File: draft-ietf-quic-http.md [hunk 12]
```diff
@@ -320,39 +449,45 @@ A proxy that supports CONNECT establishes a TCP connection ({{!RFC0793}}) to the
...
-data stream by the proxy. Note that the size and number of TCP segments is not
-guaranteed to map predictably to the size and number of QUIC STREAM frames.
-
-The TCP connection can be closed by either peer. When the client half-closes the
-data stream, the proxy will set the FIN bit on its connection to the TCP server.
-When the proxy receives a packet with the FIN bit set, it will half-close the
-corresponding data stream. TCP connections which remain half-closed in a single
+the response stream.
+
+The data received on the data streams associated with this exchange after the
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: Might want to be explicit that the request/response streams will need to have HAS_BODY, but I know that's considered vestigial.

File: draft-ietf-quic-http.md (hunk 7 above)
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: Agreed.  Any mention of a request having a control stream is leftover from my old terminology -- "on the push stream."

File: draft-ietf-quic-http.md [hunk 13]
```diff
@@ -363,23 +498,25 @@ pushes via the SETTINGS_DISABLE_PUSH setting in the SETTINGS frame (see
...
-corresponding data stream.
+The server push response is conveyed on a push stream.  Aside from the stream
+header, which identifies the response stream and PUSH_PROMISE frame, a push
+stream is identical to a regular response stream ({{stream-response}}), with
+response headers and (if present) trailers carried by HEADERS frames sent on the
+control stream, and response body (if any) indicated with a HAS_BODY frame and
+associated data stream.
 
 
 # HTTP Framing Layer
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: I think the {{http-framing-layer}} link is intended to point here, but it doesn't have an anchor.

File: draft-ietf-quic-http.md [hunk 14]
```diff
@@ -452,15 +600,26 @@ until the previous zero-value header block has been confirmed received.
...
+any other stream MUST be treated as a HTTP_WRONG_STREAM error.
 
+The format of this frame has been modified from the definition in HTTP/2 to
+accommodate not being sent on-stream and the larger stream ID space of QUIC.
 The semantics of the Stream Dependency, Weight, and E flag are the same as in
 HTTP/2.
 
 The flags defined are:
 
+  PUSH (0x04):
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: For consistency, this might be PUSH_PRIORITIZED.  Not stuck on that, though.

File: draft-ietf-quic-http.md [hunk 15]
```diff
@@ -469,33 +628,114 @@ The flags defined are:
...
+  : A 32-bit identifier for a request.  This contains the stream ID of a request
+    stream when the PUSH flag is clear, or a push ID when the PUSH flag is set.
 
   Stream Dependency:
-  : A 32-bit stream identifier for the stream that this stream depends on (see
-    {{priority}} and {!RFC7540}} Section 5.3).
+  : A 32-bit stream identifier for a dependent request.  This contains the
+    stream ID of a request stream when the PUSH flag is clear, or a push ID when
+    the PUSH flag is set.  A request stream ID of 0 indicates a dependency on
+    the root stream. For details of dependencies, see{{priority}} and
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: Missing a space.

File: draft-ietf-quic-http.md [hunk 16]
```diff
@@ -469,33 +628,114 @@ The flags defined are:
...
+The CANCEL_REQUEST frame payload has the following fields:
+
+  Request Stream/Push ID:
+  : A 32-bit stream ID or Push ID that identifies the request.  If the PUSH flag
+    is set, this refers to a Push ID included in a PUSH_PROMISE frame (see
+    {{frame-push-promise}}); if the PUSH flag is cleared, this refers to a
+    request by its stream ID.
+
+  Error Code:
+  : A 32-bit error code indicating why the response is not desired.
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: Feels a little weird lifting the QUIC error code space into the HTTP frames as well.  Are we sure we need the error code at this level?  "I no longer want it" is already implied by the frame.

File: draft-ietf-quic-http.md [hunk 17]
```diff
@@ -180,103 +180,232 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+
+An endpoint MUST terminate the connection with an HTTP_INVALID_STREAM_HEADER
+error if it receives a stream header that uses an unknown or unsupported type.
+
+
+### Stream 1: Connection Control Stream {#stream-control}
+
+Stream 1 is opened by both client and server and is used for the SETTINGS frame
+immediately after the connection opens.  After the SETTINGS frame has been sent,
+this stream is used for PRIORITY frames.
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: And CANCEL_REQUEST frames

File: draft-ietf-quic-http.md [hunk 18]
```diff
@@ -648,7 +889,9 @@ entire connection when an error is encountered.  These are referred to as
 [QUIC-TRANSPORT].
 
 HTTP/QUIC requires that only data streams be terminated abruptly.  Terminating a
-message control stream will result in an error of type HTTP_RST_CONTROL_STREAM.
+request, response or push stream will result in an error of type
+HTTP_RST_CONTROL_STREAM.  \[Editor's note: this is clearly busted right now,
+because data streams don't always exist.]
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: But not so badly, since you can always use CANCEL_REQUEST instead of telling QUIC to reset the stream.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: It's the associatives that are hard to parse, I think.  ("all data" AND ("STREAM with FIN" OR "RST_STREAM")) versus (("all data" AND "STREAM with FIN") OR "RST_STREAM")

I know which one you mean, and probably everyone in the WG does, but someone reading this document to understand QUIC might get confused.  The "whens" do delineate two cases, but I can understand that being glossed over.

**Comment by mnot** [comment]
@MikeBishop could you please summarise your support on-list, so it's not lost here?

**Comment by MikeBishop** [comment]
@mnot, that started as an e-mail, but I decided I'd better lift out the specific feedback into a GitHub review first.  ;-)  You'll see the remaining e-mail momentarily.

File: draft-ietf-quic-http.md [hunk 19]
```diff
@@ -469,33 +628,114 @@ The flags defined are:
...
+
+  Error Code:
+  : A 32-bit error code indicating why the response is not desired.
+
+A CANCEL_REQUEST frame always contains an 8 octet payload.  A server MUST treat
+a CANCEL_REQUEST frame payload that is other than 8 octets in length, or a
+CANCEL_REQUEST frame that identifies a stream that doesn't carry a valid request
+as an HTTP_MALFORMED_CANCEL_REQUEST error.  Note however that due to packet
+reordering, a CANCEL_REQUEST frame could arrive before the request stream that
+it references is opened.
```
Comment (ekr) [inline draft-ietf-quic-http.md]: This is weird text. Isn't it always the case that if you receive an unexpected frame structure you are sad?

File: draft-ietf-quic-http.md [hunk 20]
```diff
@@ -469,33 +628,114 @@ The flags defined are:
...
+### CANCEL_REQUEST {#frame-cancel-request}
+
+The CANCEL_REQUEST frame (type=0x3) is used to request cancellation of a request
+prior to the response stream being created.  This frame can be used to cancel
+both outstanding requests initiated by a client and server push requests that
+are initiated with a PUSH_PROMISE.
+
+The frame identifies a client-initiated request by the stream ID that carried
+the request; a server push request is identified by Push ID (see
+{{frame-push-promise}}).
```
Comment (ekr) [inline draft-ietf-quic-http.md]: Don't you mean "request ID" that's what you are using for dependencies

File: draft-ietf-quic-transport.md [hunk 21]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
-In the absence of more specific guidance elsewhere in this document,
-implementations SHOULD treat the receipt of a frame that is not expressly
-permitted in the description of a state as a connection error (see
-{{error-handling}}).
+The recipient of a stream will have a delayed view of the state of streams at
+its peer.  Loss or delay of frames can cause frames to arrive out of order.
 
 
-### idle
+### The "idle" State
```
Comment (huitema) [inline draft-ietf-quic-transport.md]: I am not sure whether RST_STREAM means "I won't send anymore on that stream" or "I won't receive anymore on that stream". Don't we need both?

Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: It means "I won't send," I'm fairly sure.  That's the allusion to #171, which defines a separate "I won't receive any more" frame.

File: draft-ietf-quic-transport.md [hunk 22]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
+                        receive RST_STREAM
+                                 |
+                                 v
+                            +--------+
+                            |        |
+                            | closed |
+                            |        |
+                            +--------+
+~~~
+{: #stream-state-recv title="Stream States for Receipt"}
```
Comment (huitema) [inline draft-ietf-quic-transport.md]: This new diagram shows an undocumented change in the meaning of RST_STREAM. In the current version, RST_STREAM is bilateral. "After sending a RST_STREAM, an endpoint ceases transmission of STREAM frames on
the identified stream.  A receiver of RST_STREAM can discard any data that it
already received on that stream.  An endpoint sends a RST_STREAM in response to
a RST_STREAM unless the stream is already closed." Obviously, this "send an RST back" will not work in an unilateral setup, and at a minimum that text should be updated. But if we want to keep a "right of refusal", we need a new frame, maybe "GOAWAY_STREAM", to indicate that the receiver is now discarding that stream.

File: draft-ietf-quic-transport.md [hunk 23]
```diff
@@ -2291,10 +2291,10 @@ Too Big messages.
 provisional until QUIC's loss detection algorithm determines that the packet is
```
Comment (huitema) [inline draft-ietf-quic-transport.md]: I would like to see some discussion of unilateral and GOAWAY frame. The current text describing that frame assumes bilateral streams.

File: draft-ietf-quic-transport.md [hunk 24]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
 
-Due to reordering, an endpoint could continue receiving frames for the stream
-even after the stream is closed for sending.  Frames received after a peer
-closes a stream SHOULD be discarded.  An endpoint MAY choose to limit the period
-over which it ignores frames and treat frames that arrive after this time as
-being in error.
+In the "closed" state, a sending endpoint can retransmit data that has already
+been sent in STREAM frames, or retransmit a RST_STREAM frame.  Retransmission of
+data is permitted if the packets carrying the original data are determined to
+have been lost.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: > how would this be possible if we only transition to closed state only when either:

The sender does not wait for ACK before entering closed state.

File: draft-ietf-quic-http.md (hunk 16 above)
Comment (martinthomson) [inline draft-ietf-quic-http.md]: I've removed it.  I almost didn't include this in the first place.

File: draft-ietf-quic-transport.md [hunk 25]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
+                      send STREAM with FIN flag
+                         send RST_STREAM
+                                 |
+                                 v
+                            +--------+
+                            |        |
+                            | closed |
+                            |        |
+                            +--------+
+~~~
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: @janaiyengar and I agreed to make the draft consistent about closing when the frame is first sent.  Based on reading #572, I think that we should revert that change.  The states are basically advisory only and the degree to which the documentation manages to capture these important notions makes it more useful to enclose rather than to leave retransmission up to the "transport".

File: draft-ietf-quic-transport.md (hunk 21 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I considered splitting the descriptions, but they are really still the same states.

File: draft-ietf-quic-transport.md [hunk 26]
```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
...
-Reordering might cause frames to be received after closing, see
-{{state-hc-remote}}.
+An endpoint could continue receiving frames for the stream even after the stream
+is closed for sending if packets are reordered.  At the receiving endpoint,
+retransmissions of data in STREAM frames, or retransmissions of STREAM_BLOCKED
+and RST_STREAM frames might arrive after the stream becomes "closed".  At the
+sending endpoint, MAX_STREAM_DATA frames might be received after the stream
+becomes "closed".  Frames received on a "closed" stream SHOULD be discarded.  An
+endpoint MAY choose to limit the period over which it ignores frames and treat
+frames that arrive after this time as being in error.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: If a peer closes the stream and continues to send frames on that stream long after closing, it's a way to generate unaccounted traffic.

This is a case where I would reference the ENHANCE_YOUR_CALM error, but apparently we can't have nice things.

File: draft-ietf-quic-http.md [hunk 27]
```diff
@@ -180,103 +180,232 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+After the stream header, data streams contain the octets of the payload body.
+Data streams do not carry frames.
+
+Data streams MUST reference a request or response stream.  The referenced stream
+MUST include a HAS_BODY frame.  If a data stream is received that references a
+stream of a different type, the referenced stream does not include a HAS_BODY
+frame, or another data stream already references the identified steam, then the
+connection MUST be terminated with a HTTP_UNMATCHED_BODY error.  Note that due
+to reordering of packets, the referenced stream or HAS_BODY frame might not have
+been received when the data stream is opened.
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: Yeah, you can error out earlier if you receive another header block as well in the final clause.

File: draft-ietf-quic-http.md [hunk 28]
```diff
@@ -180,103 +180,232 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+containing the message headers of informational (1xx) HTTP responses (see
+{{!RFC7230}}, Section 3.2 and {{!RFC7231}}, Section 6.2).
+
+A message that contains a payload body includes a HAS_BODY frame immediately
+after the first header block.  This indicates that another stream will include
+the payload body.  A data stream is opened to carry the payload body.  The data
+stream MUST be closed immediately after sending the body.  If the message does
+not contain a body, a data stream MUST NOT be opened and the HAS_BODY message
+MUST NOT be sent.  The "chunked" transfer encoding defined in Section 4.1 of
+{{!RFC7230}} MUST NOT be used.
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: I wanted to be able to create the body at any time, without knowing what stream might be used.  It's a minor thing, but it also keeps all pointers as back references.

File: draft-ietf-quic-http.md [hunk 29]
```diff
@@ -401,6 +538,17 @@ All frames have the following format:
...
 
+### HAS_BODY {#frame-has-body}
+
+The HAS_BODY frame (type=0x0) indicates that the HTTP message contains a payload
+body.  It is sent after the initial header block on a request, response, or push
+stream.
+
+The HAS_BODY frame has no flags and is always empty.  A HAS_BODY frame with a
+non-empty payload or flags that are set MUST be treated as a
+HTTP_MALFORMED_HAS_BODY error.
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: I considered that, and it's definitely an optimization that we can talk about later, so I went with the simplest option I could think of. The problem with the header block is that there are potentially multiple places you would put this.  The easiest approach is to grab a flag, but those appear in multiple places, so does this have to be consistent, or do you only allowing if END_HEADERS is also set, or can it appear on any HEADERS frame?

File: draft-ietf-quic-http.md (hunk 20 above)
Comment (martinthomson) [inline draft-ietf-quic-http.md]: I should make that more consistent, but I had more room here to break the two out.  I think that I'm starting to warm to Christian's notion of this needing to draw from the same 32-bit space with a bit stolen to allow discrimination rather than rely on external signals.

File: draft-ietf-quic-transport.md (hunk 23 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I'm operating here on the assumption that we are going to execute on the "application-layer shutdown" notion.  It's not at all clear to me that GOAWAY in its current form works as intended and this certainly messes up GOAWAY pretty badly.  My belief is that GOAWAY is much easier to implement at the HTTP layer where context is better understood.   You only need to concern yourself with requests and the pushes they trigger, which can simplify things a lot.  A fully generic graceful shutdown at the transport layer might not be achievable, despite the lofty goals of #604.

File: draft-ietf-quic-transport.md (hunk 22 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: This is #171, I explicitly mentioned this in the writeup.

File: draft-ietf-quic-http.md [hunk 30]
```diff
@@ -180,103 +180,231 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+
+Stream 1 is opened by both client and server and is used for the SETTINGS frame
+immediately after the connection opens.  After the SETTINGS frame has been sent,
+this stream is used for PRIORITY and CANCEL_REQUEST frames.
+
+There is no stream header for a connection control stream; the stream is
+identified by its stream number.  The connection control stream contains a
+sequence of frames, starting with the SETTINGS frame.
+
+If an endpoint receives a stream type of 0xff on any stream other than stream 1,
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: I think two versions collided here.  Stream type 0xff is undefined, so it's also invalid on Stream 1.

File: draft-ietf-quic-http.md [hunk 31]
```diff
@@ -180,103 +180,228 @@ frames, exposing the data contained within as a reliable byte stream to the
...
+~~~~~
+
+After the stream header, data streams contain the octets of the payload body.
+Data streams do not carry frames.
+
+Data streams MUST reference a request or response stream.  The referenced stream
+MUST include a HAS_BODY frame.  If a data stream is received that references a
+stream of a different type, the referenced stream does not include a HAS_BODY
+frame, or another data stream already references the identified steam, then the
+connection MUST be terminated with a HTTP_UNMATCHED_BODY error.  Note that due
```
Comment (MikeBishop) [inline draft-ietf-quic-http.md]: HTTP_UNMATCHED_DATA
